daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1481,7 +1487,12 @@ private[lf] object SBuiltin {
                 onLedger.ptx = onLedger.ptx.copy(contractState = next)
                 keyMapping match {
                   case ContractStateMachine.KeyActive(cid) =>
-                    operation.handleKeyFound(machine, cid)
+                    if (onLedger.cachedContracts.contains(cid)) {
+                      ensureVisible(cid)
+                    } else {
+                      machine.pushKont(KContinue(_ => ensureVisible(cid)))
+                      machine.ctrl = SBFetchAny(SEValue(SContractId(cid)), SBSome(SEValue(skey)))
```

Comment:
maybe add a comment here that this adds it to cachedContracts. That makes it a bit easier to understand why `ensureVisible` is going to succeed the next time.